python/ray/train/lightning/lightning_utils.py

```diff
@@ -0,0 +1,117 @@
+import os
+import logging
+import torch
+from torch import Tensor
+from typing import Any, Dict, Optional
+from copy import deepcopy
+
+import pytorch_lightning as pl
+from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.utilities import rank_zero_info, rank_zero_only
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+from pytorch_lightning.strategies import DDPStrategy
+from lightning_fabric.plugins.environments.lightning import LightningEnvironment
+
+import ray
+from ray.air import session
+from ray.train.lightning.lightning_checkpoint import LightningCheckpoint
+
+
+logger = logging.getLogger(__name__)
+
+class RayModelCheckpoint(ModelCheckpoint):
+    """AIR customized ModelCheckpoint callback """
+
+    def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: str) -> None:
+        super().setup(trainer, pl_module, stage)
+        self.last_best_k_models = {}
+
+    def format_checkpoint_name(
+        self, metrics: Dict[str, Tensor], filename: Optional[str] = None, ver: Optional[int] = None
+    ) -> str:
+        """
+        Ensure different checkpoint files saved in seperate folders to align with AIR checkpoint format.
+
+        e.g. './epoch=2-validation_loss=0.12.ckpt' -> './epoch=2-validation_loss=0.12/checkpoint.ckpt'
+        """
+        filepath = super().format_checkpoint_name(metrics, filename, ver)
+        filepath = filepath.replace(self.FILE_EXTENSION, f"/checkpoint{self.FILE_EXTENSION}")
+        return filepath
+
+    def _session_report(self, trainer: "pl.Trainer"):
+        """Report latest metrics dict and checkpoint to AIR training session."""
+        kwargs = {}
+
+        metrics = self._monitor_candidates(trainer)
+        for k, v in metrics.items():
+            if isinstance(v, torch.Tensor):
+                metrics[k] = v.cpu().numpy()
+        kwargs["metrics"] = metrics
+
+        new_checkpoint = self.best_k_models.keys() - self.last_best_k_models.keys()
+        if len(new_checkpoint) > 1:
+            logger.warning(
+                "Multiple checkpoints were created between two training steps, but only one of them will be reported."
+            )
+
+        if len(new_checkpoint) == 1:
+            filepath = new_checkpoint.pop()
+            if trainer.global_rank == 0:
+                kwargs["checkpoint"] = LightningCheckpoint.from_directory(path=os.path.dirname(filepath))
+            else:
+                kwargs["checkpoint"] = LightningCheckpoint.from_dict({"rank": session.get_world_rank()})
```

Comment:
How does the 2nd branch work? Is it just some dummy data since we only take rank 0 checkpoint?

Reply:
Yes. If we don't pass a dummy checkpoint to the other workers, the program will hang forever.

Reply:
Got it, can we make a comment about that?